sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -2364,11 +2364,20 @@ class Analyzer(override val catalogManager: CatalogManager)
      * Resolve and rewrite all subqueries in an operator tree..
      */
     def apply(plan: LogicalPlan): LogicalPlan = plan.resolveOperatorsUpWithPruning(
-      _.containsAnyPattern(SCALAR_SUBQUERY, EXISTS_SUBQUERY, IN_SUBQUERY), ruleId) {
+      _.containsPattern(PLAN_EXPRESSION), ruleId) {
       // In case of HAVING (a filter after an aggregate) we use both the aggregate and
       // its child for resolution.
       case f @ Filter(_, a: Aggregate) if f.childrenResolved =>
         resolveSubQueries(f, Seq(a, a.child))
+      case j @ LateralJoin(left, right, _, condition) if left.resolved && !right.resolved =>
+        val newRight = resolveSubQuery(right, j.children)(LateralSubquery(_, _, right.exprId))
+          .asInstanceOf[LateralSubquery]
+        val newCond = if (newRight.resolved && condition.exists(!_.resolved)) {
+          condition.map(resolveExpressionByPlanOutput(_, newRight.plan))
```

Comment:
It's ok to have special column resolution logic. If you checkout the rule `ResolveReferences`, there are many special cases.

Let's put the special case for `LateralJoin` in `ResolveReferences` as well, to centralize all the column resolution logic. Let's also follow the general column resolution rule: fail if the name is ambiguous. For example, if join condition refers to a column in both left and right sides (un-qualified column name), we should fail.